# Example 4.2
在本例題中，要探討Michigan Educational Assessment Program(MEAP)standardized 10th grade math test(math10)與其他變數的互動
## R-square 判斷法

In [40]:
import os
import pandas as pd
import statsmodels.api as sm
import numpy as np

In [3]:
path = os.path.abspath(r'C:\Users\s0954\Desktop\財金計量方法\0521複回歸作業\複迴歸結果解釋_Wooldrige_chapter_4_樣本資料-20210523')
file = 'Example4.2_Meap93.csv'
df = pd.read_csv(os.path.join(path,file))

In [4]:
df.head(10)

lnchprg  enroll       staff  expend  salary  benefits  droprate  \
0      1.4    1862  112.599999    5765   37498      7420       2.9   
1      2.3   11355  101.199997    6601   48722     10370       1.3   
2      2.7    7685  114.000000    6834   44541      7313       3.5   
3      3.4    1148   85.400002    3586   31566      5989       3.6   
4      3.4    1572   96.099998    3847   29781      5545       0.0   
5      3.4    2496  101.099998    5070   36801      5895       2.7   
6      3.6    3358   87.599998    4474   37863      6934       1.4   
7      3.6   11983   96.000000    5159   40133      8085       2.5   
8      4.2    3499  102.800003    5012   36451      7253       1.8   
9      4.2    5095  145.000000    6501   33449      7386       1.9   

     gradrate     math10      sci11  totcomp   ltotcomp   lexpend   lenroll  \
0   89.199997  56.400002  67.900002    44918  10.712594  8.659560  7.529407   
1   91.400002  42.700001  65.300003    59092  10.986851  8.794976  9.337414   
2   91.400002  43.799999  54.299999    51854  10.856188  8.829665  8.947025   
3   86.599998  25.299999  60.000000    37555  10.533562  8.184793  7.045776   
4  100.000000  15.300000  65.800003    35326  10.472375  8.255049  7.360104   
5   89.199997  46.000000  60.500000    42696  10.661860  8.531096  7.822445   
6   95.099998  33.599998  67.400002    44797  10.709896  8.406038  8.119101   
7   87.800003  40.099998  69.400002    48218  10.783487  8.548498  9.391244   
8   93.199997  42.099998  71.699997    43704  10.685195  8.519590  8.160233   
9   92.099998  39.799999  55.000000    40835  10.617295  8.779712  8.536015   

     lstaff    bensal    lsalary  
0  4.723842  0.197877  10.532043  
1  4.617099  0.212840  10.793886  
2  4.736198  0.164186  10.704165  
3  4.447346  0.189729  10.359836  
4  4.565389  0.186193  10.301626  
5  4.616110  0.160186  10.513280  
6  4.472781  0.183134  10.541730  
7  4.564348  0.201455  10.599955  
8  4.632785  0.198979  10.503724  
9  4.976734  0.220814  10.417777

In [5]:
math10 = df['math10']
del df['math10']

## 每一項跑單迴歸分析

In [6]:
#計算每個自變數 X 的 pvalue
def single_pval(X,y):
    pv = {}
    for single_col in X:
        model = sm.OLS(y,sm.add_constant(df[single_col])).fit()
        single_pvalue = model.pvalues.iloc[1]
        pv[single_col]=single_pvalue
    return pv

In [7]:
def single_rsq(X,y):
    rsq = {}
    for single_col in X:
        model = sm.OLS(y,sm.add_constant(df[single_col])).fit()
        single_rsq = model.rsquared
        rsq[single_col]=single_rsq
    return rsq

In [8]:
pv = single_pval(df,math10)
pv
#pd.DataFrame.from_dict(pv,columns=['pvalue'])

C:\Users\s0954\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


{'lnchprg': 2.746609039618981e-18,
 'enroll': 0.14827081751261023,
 'staff': 0.9071676822326178,
 'expend': 0.0002273004741148899,
 'salary': 1.02785910421596e-06,
 'benefits': 0.08686033124794995,
 'droprate': 0.00025452723970975383,
 'gradrate': 4.747924361105861e-07,
 'sci11': 3.978634773844495e-06,
 'totcomp': 6.013991871100513e-06,
 'ltotcomp': 1.7450013302877949e-06,
 'lexpend': 0.00047520715083538775,
 'lenroll': 0.3277926655003991,
 'lstaff': 0.6980739767952057,
 'bensal': 0.06804197358659328,
 'lsalary': 4.111765460216469e-07}

In [9]:
rsq = single_rsq(df,math10)
rsq

{'lnchprg': 0.1710340964985384,
 'enroll': 0.005140744341830428,
 'staff': 3.3534093119236275e-05,
 'expend': 0.03296049962828518,
 'salary': 0.05717830007027236,
 'benefits': 0.007203558890136574,
 'droprate': 0.03245411405482801,
 'gradrate': 0.06063250636268258,
 'sci11': 0.051113462626346196,
 'totcomp': 0.049259834610870246,
 'ltotcomp': 0.05480820724123947,
 'lexpend': 0.02966329326596473,
 'lenroll': 0.0023587368150796983,
 'lstaff': 0.00037103627207901724,
 'bensal': 0.008177550936759315,
 'lsalary': 0.06127523769013998}

In [10]:
pv_df = pd.DataFrame(pv.items(),columns=['name','pvalue']).set_index('name')
rsq_df = pd.DataFrame(rsq.items(),columns=['name','rsquared']).set_index('name')
variable_df =pd.concat([pv_df,rsq_df],axis=1) 

In [11]:
variable_df = variable_df[variable_df.index!='math10']

In [12]:
#variable_df.applymap("{0:.3f}".format)
variable_df

pvalue  rsquared
name                            
lnchprg   2.746609e-18  0.171034
enroll    1.482708e-01  0.005141
staff     9.071677e-01  0.000034
expend    2.273005e-04  0.032960
salary    1.027859e-06  0.057178
benefits  8.686033e-02  0.007204
droprate  2.545272e-04  0.032454
gradrate  4.747924e-07  0.060633
sci11     3.978635e-06  0.051113
totcomp   6.013992e-06  0.049260
ltotcomp  1.745001e-06  0.054808
lexpend   4.752072e-04  0.029663
lenroll   3.277927e-01  0.002359
lstaff    6.980740e-01  0.000371
bensal    6.804197e-02  0.008178
lsalary   4.111765e-07  0.061275

In [13]:
def regression(X,y):
    return sm.OLS(y,sm.add_constant(X)).fit()

## 開始篩選
先看全丟進去的

In [14]:
print(regression(df,math10).summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.277
Model:                            OLS   Adj. R-squared:                  0.249
Method:                 Least Squares   F-statistic:                     10.02
Date:                Tue, 25 May 2021   Prob (F-statistic):           2.99e-20
Time:                        01:41:51   Log-Likelihood:                -1471.3
No. Observations:                 408   AIC:                             2975.
Df Residuals:                     392   BIC:                             3039.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -368.6624    323.277     -1.140      0.2

## 先挑p-value顯著&R-square高的

In [15]:
mask1 = variable_df['pvalue']<0.05
mask2 = variable_df['rsquared']>0.01

In [16]:
selected = variable_df[mask1&mask2].index
variable_df[mask1&mask2]

pvalue  rsquared
name                            
lnchprg   2.746609e-18  0.171034
expend    2.273005e-04  0.032960
salary    1.027859e-06  0.057178
droprate  2.545272e-04  0.032454
gradrate  4.747924e-07  0.060633
sci11     3.978635e-06  0.051113
totcomp   6.013992e-06  0.049260
ltotcomp  1.745001e-06  0.054808
lexpend   4.752072e-04  0.029663
lsalary   4.111765e-07  0.061275

In [17]:
result1 = regression(df[selected],math10) 
print(result1.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.236
Model:                            OLS   Adj. R-squared:                  0.217
Method:                 Least Squares   F-statistic:                     12.29
Date:                Tue, 25 May 2021   Prob (F-statistic):           1.28e-18
Time:                        01:41:51   Log-Likelihood:                -1482.5
No. Observations:                 408   AIC:                             2987.
Df Residuals:                     397   BIC:                             3031.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -190.5173    300.977     -0.633      0.5

再挑出 P-value<0.05的

In [18]:
selected2 = result1.pvalues[result1.pvalues<0.05].index

In [19]:
result2 = regression(df[selected2],math10)

In [20]:
print (result2.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.226
Model:                            OLS   Adj. R-squared:                  0.213
Method:                 Least Squares   F-statistic:                     16.70
Date:                Tue, 25 May 2021   Prob (F-statistic):           2.37e-19
Time:                        01:41:52   Log-Likelihood:                -1485.2
No. Observations:                 408   AIC:                             2986.
Df Residuals:                     400   BIC:                             3019.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -93.6761    298.418     -0.314      0.7

再篩一次 P-value<0.05

In [21]:
selected3 = result2.pvalues[result2.pvalues<0.05].index
result3 = regression(df[selected3],math10)
print(result3.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.226
Model:                            OLS   Adj. R-squared:                  0.214
Method:                 Least Squares   F-statistic:                     19.47
Date:                Tue, 25 May 2021   Prob (F-statistic):           5.87e-20
Time:                        01:41:52   Log-Likelihood:                -1485.4
No. Observations:                 408   AIC:                             2985.
Df Residuals:                     401   BIC:                             3013.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -98.7071    298.008     -0.331      0.7

## 僅用R-square篩選

In [22]:
setA = set(variable_df[variable_df['rsquared']>0.01].index)
setB = set(variable_df[variable_df['rsquared']>0.001].index)
setB-setA

{'benefits', 'bensal', 'enroll', 'lenroll'}

In [23]:
selected_1a = variable_df[variable_df['rsquared']>0.01].index
result_1a = regression(df[selected_1a],math10)
print (result_1a.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.236
Model:                            OLS   Adj. R-squared:                  0.217
Method:                 Least Squares   F-statistic:                     12.29
Date:                Tue, 25 May 2021   Prob (F-statistic):           1.28e-18
Time:                        01:41:52   Log-Likelihood:                -1482.5
No. Observations:                 408   AIC:                             2987.
Df Residuals:                     397   BIC:                             3031.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -190.5173    300.977     -0.633      0.5

In [24]:
selected_1b = variable_df[variable_df['rsquared']>0.001].index
result_1b = regression(df[selected_1b],math10)
print (result_1b.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.271
Model:                            OLS   Adj. R-squared:                  0.246
Method:                 Least Squares   F-statistic:                     11.24
Date:                Tue, 25 May 2021   Prob (F-statistic):           1.33e-20
Time:                        01:41:52   Log-Likelihood:                -1473.2
No. Observations:                 408   AIC:                             2974.
Df Residuals:                     394   BIC:                             3031.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -360.7214    297.992     -1.211      0.2

結論:
若使用單一變數X的p-value顯著程度與其R-squared來去捨，
模型明顯會遺失一些解釋能力($0.277\rightarrow0.236\rightarrow0.226$)；

而若只考慮R-squared較低的變數被移除之後({'benefits', 'bensal', 'enroll', 'lenroll'})，
還是可以發現他們影響力不小($0.271\rightarrow0.236$)

因此，需更進一步去探討OVB、Multicollinearity、Quadratics去挑選自變數

## OVB 舉例
由於自變數項很多，挑兩個p-value顯著的來測試，以 expand,salary為例

In [25]:
expend = df['expend']
salary = df['salary']
X = pd.concat([expend,salary],axis=1)
y = math10

### $ y = \beta_0 + \beta_1 x_1+\beta_2 x_2+u $
### $ y = \beta_0 + \tilde{\beta_1} x_1+u $
### $ x_2 = \delta_0+\delta_1 x_1+v $
### $ y = \beta_0 + \beta_1 x_1+\color{red}{\beta_2 (\delta_0+\delta_1 x_1+v)}+u $

### $ \tilde{\beta_1} = \hat{\beta_1}+\hat{\beta_2}\tilde{\delta_1}$
### $ \hat{\beta_2}\tilde{\delta_1} \rightarrow bias$


In [26]:
model = regression(X=X,y=y)
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.058
Model:                            OLS   Adj. R-squared:                  0.054
Method:                 Least Squares   F-statistic:                     12.55
Date:                Tue, 25 May 2021   Prob (F-statistic):           5.15e-06
Time:                        01:41:53   Log-Likelihood:                -1525.3
No. Observations:                 408   AIC:                             3057.
Df Residuals:                     405   BIC:                             3069.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.5525      3.390      2.228      0.0

In [27]:
ovb_1 = regression(X=salary,y = expend)

In [28]:
print(ovb_1.summary())

                            OLS Regression Results                            
Dep. Variable:                 expend   R-squared:                       0.423
Model:                            OLS   Adj. R-squared:                  0.421
Method:                 Least Squares   F-statistic:                     297.5
Date:                Tue, 25 May 2021   Prob (F-statistic):           2.09e-50
Time:                        01:41:53   Log-Likelihood:                -3181.1
No. Observations:                 408   AIC:                             6366.
Df Residuals:                     406   BIC:                             6374.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1194.9492    186.761      6.398      0.0

### $\tilde{\delta_1}=0.1001$

這邊可以看到 warnings 裡有說，可能有 multicollinearity的問題，而expend不顯著，可能自變數挑不夠好

In [29]:
sci11 = df['sci11']
totcomp = df['totcomp']
X = pd.concat([sci11,totcomp],axis=1)

In [30]:
print(regression(X=X,y=y).summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.088
Model:                            OLS   Adj. R-squared:                  0.084
Method:                 Least Squares   F-statistic:                     19.60
Date:                Tue, 25 May 2021   Prob (F-statistic):           7.51e-09
Time:                        01:41:53   Log-Likelihood:                -1518.7
No. Observations:                 408   AIC:                             3043.
Df Residuals:                     405   BIC:                             3055.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.8504      3.566      0.799      0.4

雖然還是有warnings，但至少兩個自變數都顯著

$math10 = 2.8504+0.0003\space totcomp+0.1670\space sci11$

In [31]:
ovb_2 = regression(X=totcomp,y=sci11)

In [32]:
print(ovb_2.summary())

                            OLS Regression Results                            
Dep. Variable:                  sci11   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.016
Method:                 Least Squares   F-statistic:                     7.822
Date:                Tue, 25 May 2021   Prob (F-statistic):            0.00541
Time:                        01:41:54   Log-Likelihood:                -1605.8
No. Observations:                 408   AIC:                             3216.
Df Residuals:                     406   BIC:                             3224.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         38.1816      3.981      9.590      0.0

$sci11 = 38.1816+0.0003\space totcomp $

### $\tilde{\delta_1}=0.0003$

In [33]:
model = regression(X=totcomp,y=math10)
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.049
Model:                            OLS   Adj. R-squared:                  0.047
Method:                 Least Squares   F-statistic:                     21.04
Date:                Tue, 25 May 2021   Prob (F-statistic):           6.01e-06
Time:                        01:41:54   Log-Likelihood:                -1527.2
No. Observations:                 408   AIC:                             3058.
Df Residuals:                     406   BIC:                             3066.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.2271      3.284      2.810      0.0

$math10=9.2271+0.0004\space totcomp$

### $ \tilde{\beta_1} = \hat{\beta_1}+\hat{\beta_2}\tilde{\delta_1}$

### $ \widetilde{\beta^{totcomp}} = \widehat{\beta^{totcomp}}+\widehat{\beta^{sci11}}\tilde{\delta_1}$
$0.0004 = 0.0003 + 0.1670\times0.0003$

In [34]:
0.0003 + 0.1670*0.0003

0.0003501

## Multicolinearity & VIF

這邊挑 expend,salary,droprate 為例

In [35]:
expend = df['expend']
salary = df['salary']
droprate = df['droprate']
#math10 = df['math10']
X = pd.concat([expend,salary,droprate],axis=1)
y = math10

In [36]:
model1 = regression(salary,y)
model2 = regression(X,y)
print(model1.summary())
print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.057
Model:                            OLS   Adj. R-squared:                  0.055
Method:                 Least Squares   F-statistic:                     24.62
Date:                Tue, 25 May 2021   Prob (F-statistic):           1.03e-06
Time:                        01:41:54   Log-Likelihood:                -1525.5
No. Observations:                 408   AIC:                             3055.
Df Residuals:                     406   BIC:                             3063.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          8.2822      3.229      2.565      0.0

$salary:0.0005\rightarrow0.0004$

In [37]:
X.corr()

expend    salary  droprate
expend    1.000000  0.650296  0.002288
salary    0.650296  1.000000 -0.072925
droprate  0.002288 -0.072925  1.000000

### 計算個別VIF

In [38]:
expend_VIF = 1/(1-regression(expend,math10).rsquared)
salary_VIF = 1/(1-regression(salary,math10).rsquared)
droprate_VIF = 1/(1-regression(droprate,math10).rsquared)
print(' expend_VIF:',expend_VIF,'\n','salary_VIF:',salary_VIF,'\n','droprate_VIF:',droprate_VIF)

 expend_VIF: 1.0340839227514653 
 salary_VIF: 1.0606459313298942 
 droprate_VIF: 1.033542713091198


expend跟droprate的VIF並沒有>10，也就是無須移除自變數expend,droprate

## Quadratic
用 salary, droprate, droprate^2測試

In [45]:
print (regression(pd.concat([salary,droprate],axis=1),math10).summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.084
Model:                            OLS   Adj. R-squared:                  0.079
Method:                 Least Squares   F-statistic:                     18.52
Date:                Tue, 25 May 2021   Prob (F-statistic):           2.01e-08
Time:                        01:43:28   Log-Likelihood:                -1519.7
No. Observations:                 408   AIC:                             3045.
Df Residuals:                     405   BIC:                             3057.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.6572      3.261      3.268      0.0

In [61]:
print (regression(pd.concat([salary,droprate,np.power(droprate,2).rename('droprate^2')],axis=1),math10).summary())

                            OLS Regression Results                            
Dep. Variable:                 math10   R-squared:                       0.135
Model:                            OLS   Adj. R-squared:                  0.128
Method:                 Least Squares   F-statistic:                     20.93
Date:                Tue, 25 May 2021   Prob (F-statistic):           1.26e-12
Time:                        01:49:51   Log-Likelihood:                -1508.1
No. Observations:                 408   AIC:                             3024.
Df Residuals:                     404   BIC:                             3040.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.5349      3.228      4.193      0.0

$R-Squared: \space 0.084\rightarrow 0.135$

$math10 = 10.6572 + 0.0005\space salary - 0.3130 \space droprate      $

$math10 = 13.5349 + 0.0005\space salary - 0.9696 \space droprate + 0.0197droprate^2      $

### $\frac{\Delta math10}{\Delta droprate} = -0.9696 + 2(0.0197)droprate$

## Stepwise Regression

### 向前法 forward

In [181]:
def forward_regression(X,y,critcal_p):
    selected = []
    iteration = 0
    while True:
        print('--------------iteration',str(iteration),'----------------')
        changed = False
        
        #還未被選入的自變數
        unselected = list(set(X)-set(selected))
        print('selected',selected)
        print('unselected:',unselected)
        
        #每一次迭代的pvalue
        #格式為 col_name: pvalue, e.g. lnchprg : 2.746609039618981e-18
        pv={}
        
        #從未被選入的自變數一次一個加入計算pvalue
        #e.g.  selected :a,b,c , unselected:d,e,f , iter: abcd,abce,abcf
        for i in unselected:
            pv[i]= regression(pd.concat([X[selected],X[i]],axis=1),y).pvalues.iloc[-1]  #新加入的在最後一個，index = -1
            #print(pv.items())
        smallest_p = min(pv.items(), key=lambda x: x[1]) #這邊是 key-value的形式，比較的時候要用 value比
        print(smallest_p[0],':',smallest_p[1]) # x[0]=col_name,x[1]=pvalue
        if smallest_p[1] < critcal_p: #判斷有沒有超過臨界值
            selected.append(smallest_p[0])
            print('select:',smallest_p[0])
            changed = True
        else:
            print('do not select:',smallest_p[0])
        if changed==False:
            print('--------------    end    ----------------')
            break
        iteration = iteration+1
    return selected

In [226]:
f_selected = forward_regression(df,math10,0.05)
print(regression(df[f_selected],math10).summary())

--------------iteration 0 ----------------
selected []
unselected: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend', 'droprate']
lnchprg : 2.746609039618981e-18
select: lnchprg
--------------iteration 1 ----------------
selected ['lnchprg']
unselected: ['enroll', 'salary', 'lenroll', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend', 'droprate']
gradrate : 0.007008764108634065
select: gradrate
--------------iteration 2 ----------------
selected ['lnchprg', 'gradrate']
unselected: ['enroll', 'salary', 'lenroll', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'sci11', 'bensal', 'expend', 'droprate']
expend : 0.014688604564107706
select: expend
--------------iteration 3 ----------------
selected ['lnchprg', 'gradrate', 'expend']
unselected: ['enroll', 'salary', 'lenroll', 'lsalary', 'ben

### 向後法 backward 

In [230]:
def backward_regression(X,y,critcal_p):
    excluded = []
    iteration = 0
    while True:
        print('--------------iteration',str(iteration),'----------------')
        change=False
        pv={}
        included = list(set(X)-set(excluded))
        print('excluded:',excluded)
        print('included:',included)
        pv = dict(regression(X[included],y).pvalues[1:])  #不要const
        max_p = max(pv.items(),key=lambda x:x[1])
        print (max_p[0],':',max_p[1])
        if max_p[1] > critcal_p:
            excluded.append(max_p[0])
            change=True
            print('exclude:',max_p[0])
        else:
            print('do not exclude:',max_p[0])
        if change == False:
            break
        iteration = iteration+1
    return list(set(X)-set(excluded))

In [231]:
b_selected = backward_regression(df,math10,0.05)
print(regression(df[b_selected],math10).summary())

--------------iteration 0 ----------------
excluded: []
included: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend', 'droprate']
droprate : 0.7223388710430864
exclude: droprate
--------------iteration 1 ----------------
excluded: ['droprate']
included: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend']
lstaff : 0.5239365578473953
exclude: lstaff
--------------iteration 2 ----------------
excluded: ['droprate', 'lstaff']
included: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'gradrate', 'sci11', 'bensal', 'expend']
sci11 : 0.40058048755566833
exclude: sci11
--------------iteration 3 ----------------
excluded: ['droprate', 'lstaff', 'sci11']
included: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benef

### 逐步法 stepwise 

In [275]:
def stepwise_regression(X,y,critcal_p):
    selected = []
    iteration = 0
    while True:
        print('--------------iteration',str(iteration),'----------------')
        changed = False
        
        #還未被選入的自變數
        unselected = list(set(X)-set(selected))
        print('selected',selected)
        print('unselected:',unselected)
        
        #每一次迭代的pvalue
        #格式為 col_name: pvalue, e.g. lnchprg : 2.746609039618981e-18
        pv={}
        
        #從未被選入的自變數一次一個加入計算pvalue
        #e.g.  selected :a,b,c , unselected:d,e,f , iter: abcd,abce,abcf
        for i in unselected:
            pv[i]= regression(pd.concat([X[selected],X[i]],axis=1),y).pvalues.iloc[-1]  #新加入的在最後一個，index = -1
            #print(pv.items())
        smallest_p = min(pv.items(), key=lambda x: x[1]) #這邊是 key-value的形式，比較的時候要用 value比
        print(smallest_p[0],':',smallest_p[1]) # x[0]=col_name,x[1]=pvalue
        if smallest_p[1] < critcal_p: #判斷有沒有超過臨界值
            selected.append(smallest_p[0])
            print('select:',smallest_p[0])
            changed = True
        else:
            print('do not select:',smallest_p[0])
            
        #如果新自變數加入而導致有一自變數 pvalue>0.05要記得移除
        allpv = regression(X[selected],y).pvalues[1:] #不要const
        curr_p = dict(allpv)
        print('curr_p',curr_p) #現在有被選到的自變數對應的 pvalue
        curr_max_p = max(dict(curr_p).items(),key=lambda x:x[1])
        print('curr_max_p',curr_max_p)
        if curr_max_p[1] > critcal_p: #如果最不顯著者大於臨界值就要請他走
            selected.remove(curr_max_p[0])
            print('remove:',curr_max_p[0]) 
            changed = True

        if changed==False:
            print('--------------    end    ----------------')
            break
        iteration = iteration+1
    return selected


In [277]:
print(regression(df[stepwise_regression(df,math10,0.05)],math10).summary())

--------------iteration 0 ----------------
selected []
unselected: ['enroll', 'salary', 'lenroll', 'lnchprg', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend', 'droprate']
lnchprg : 2.746609039618981e-18
select: lnchprg
curr_p {'lnchprg': 2.746609039618981e-18}
curr_max_p ('lnchprg', 2.746609039618981e-18)
--------------iteration 1 ----------------
selected ['lnchprg']
unselected: ['enroll', 'salary', 'lenroll', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'gradrate', 'sci11', 'bensal', 'expend', 'droprate']
gradrate : 0.007008764108634065
select: gradrate
curr_p {'lnchprg': 2.8356059825772488e-14, 'gradrate': 0.007008764108634065}
curr_max_p ('gradrate', 0.007008764108634065)
--------------iteration 2 ----------------
selected ['lnchprg', 'gradrate']
unselected: ['enroll', 'salary', 'lenroll', 'lsalary', 'benefits', 'totcomp', 'lexpend', 'staff', 'ltotcomp', 'lstaff', 'sci11', 'bensal', '